// StreamUnary RPCs that contains streaming strings
// https://ably.com/blog/grpc-stream-performance
syntax = "proto3";

package unary;

service QueryService{
  rpc Handshake(HandshakeRequest) returns (HandshakeResponse) {}
  rpc Encode(stream EncodeRequest) returns (stream EncodeResponse) {}
  rpc Run(stream RunRequest) returns (stream RunResponse) {}
}

message HandshakeRequest {
  string client_id = 1;
}

message HandshakeResponse {
  string session_id = 1;
}

message RunRequest {
  string message = 1;
  optional string candidate = 2;
  optional string kwargs = 3;
}

message RunResponse {
  string message = 1;
  bool ranked = 2;
  bool success = 3;
  optional string logit = 4;
  optional bytes embedding = 5;
}

message EncodeRequest {
  string message = 1;
  string reduction = 3;
  optional string kwargs = 4;
}

message EncodeResponse {
  bytes embedding = 1;
  bool success = 2;
}